File: src/main/java/no/nav/sosialhjelp/soknad/consumer/organisasjon/OrganisasjonService.java
package no.nav.sosialhjelp.soknad.consumer.organisasjon;

import no.nav.sosialhjelp.soknad.client.organisasjon.OrganisasjonClient;
import no.nav.sosialhjelp.soknad.client.organisasjon.dto.NavnDto;
import no.nav.sosialhjelp.soknad.client.organisasjon.dto.OrganisasjonNoekkelinfoDto;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;

@Component
public class OrganisasjonService {

    private static final Logger log = LoggerFactory.getLogger(OrganisasjonService.class);

    private final OrganisasjonClient organisasjonClient;

    public OrganisasjonService(OrganisasjonClient organisasjonClient) {
        this.organisasjonClient = organisasjonClient;
    }

    public String hentOrgNavn(String orgnr) {
        if (orgnr != null) {
            try {
                OrganisasjonNoekkelinfoDto noekkelinfo = organisasjonClient.hentOrganisasjonNoekkelinfo(orgnr);
                if (noekkelinfo == null) {
                    log.warn("Kunne ikke hente orgnr fra Ereg: {}", orgnr);
                    return orgnr;
                }
                NavnDto navn = noekkelinfo.getNavn();
                List<String> list = new ArrayList<>(asList(navn.getNavnelinje1(), navn.getNavnelinje2(), navn.getNavnelinje3(), navn.getNavnelinje4(), navn.getNavnelinje5()));
                list.removeAll(asList("", null)); // fjern tomme strenger og null (håndteres som "null")
                return String.join(", ", list);
            } catch (Exception e) {
                log.warn("Kunne ikke hente orgnr fra Ereg: {}", orgnr, e);
                return orgnr;
            }
        } else {
            return "";
        }
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/organisasjon/OrganisasjonService.kt
package no.nav.sosialhjelp.soknad.organisasjon

import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOrganisasjon
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
open class OrganisasjonService(
    private val organisasjonClient: OrganisasjonClient
) {

    open fun hentOrgNavn(orgnr: String?): String {
        return if (orgnr != null) {
            try {
                val noekkelinfo = organisasjonClient.hentOrganisasjonNoekkelinfo(orgnr)
                if (noekkelinfo == null) {
                    log.warn("Kunne ikke hente orgnr fra Ereg: $orgnr")
                    return orgnr
                }
                mutableListOf(
                    noekkelinfo.navn.navnelinje1,
                    noekkelinfo.navn.navnelinje2,
                    noekkelinfo.navn.navnelinje3,
                    noekkelinfo.navn.navnelinje4,
                    noekkelinfo.navn.navnelinje5
                )
                    .filterNotNull()
                    .filter { it.isNotEmpty() }
                    .joinToString(separator = ", ") { it }
            } catch (e: Exception) {
                log.warn("Kunne ikke hente orgnr fra Ereg: $orgnr", e)
                orgnr
            }
        } else {
            ""
        }
    }

    fun mapToJsonOrganisasjon(orgnr: String): JsonOrganisasjon? {
        if (orgnr.matches(Regex("\\d{9}"))) {
            return JsonOrganisasjon()
                .withNavn(hentOrgNavn(orgnr))
                .withOrganisasjonsnummer(orgnr)
        }
        if (orgnr.matches(Regex("\\d{11}"))) {
            log.info("Utbetalingens opplysningspliktigId er et personnummer. Dette blir ikke inkludert i soknad.json")
        } else {
            log.error("Utbetalingens opplysningspliktigId er verken et organisasjonsnummer eller personnummer: $orgnr. Kontakt skatteetaten.")
        }
        return null
    }

    companion object {
        private val log = LoggerFactory.getLogger(OrganisasjonService::class.java)
    }
}
